commit 09615b0190ce348d9da8b65c494c9bb0f6641321 (refs/changes/07/3307/3)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-02-07T15:27:28+00:00 (1 year, 8 months ago)
    
    OS-6588 zfs_zone IO throttle assertion failure
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>
    Approved by: Mike Gerdts <mike.gerdts@joyent.com>

diff --git a/usr/src/uts/common/fs/zfs/zfs_zone.c b/usr/src/uts/common/fs/zfs/zfs_zone.c
index 59357cbee5..f151595095 100644
--- a/usr/src/uts/common/fs/zfs/zfs_zone.c
+++ b/usr/src/uts/common/fs/zfs/zfs_zone.c
@@ -654,11 +654,12 @@ zfs_zone_wait_adjust_calculate_cb(zone_t *zonep, void *arg)
 	zone_persist_t *zpd = &zone_pdata[zonep->zone_id];
 	zone_zfs_io_t *iop = zpd->zpers_zfsp;
 
-	ASSERT(MUTEX_HELD(&zpd->zpers_zfs_lock));
 	ASSERT3P(iop, !=, NULL);
 
+	mutex_enter(&zpd->zpers_zfs_lock);
 	if (zonep->zone_id == GLOBAL_ZONEID ||
 	    get_zone_io_cnt(sp->zi_now, iop, &rops, &wops, &lwops) == 0) {
+		mutex_exit(&zpd->zpers_zfs_lock);
 		return (0);
 	}
 
@@ -685,6 +686,8 @@ zfs_zone_wait_adjust_calculate_cb(zone_t *zonep, void *arg)
 	    uint_t, rops, uint_t, wops, uint_t, lwops,
 	    uint64_t, iop->zpers_io_util, uint16_t, iop->zpers_zfs_io_pri);
 
+	mutex_exit(&zpd->zpers_zfs_lock);
+
 	return (0);
 }
 
@@ -719,9 +722,9 @@ zfs_zone_wait_adjust_delay_cb(zone_t *zonep, void *arg)
 	uint8_t delay;
 	uint_t fairutil = 0;
 
-	ASSERT(MUTEX_HELD(&zpd->zpers_zfs_lock));
 	ASSERT3P(iop, !=, NULL);
 
+	mutex_enter(&zpd->zpers_zfs_lock);
 	delay = iop->zpers_io_delay;
 	iop->zpers_io_util_above_avg = 0;
 
@@ -763,6 +766,8 @@ zfs_zone_wait_adjust_delay_cb(zone_t *zonep, void *arg)
 	    uintptr_t, delay, uintptr_t, iop->zpers_io_delay,
 	    uintptr_t, fairutil, uintptr_t, iop->zpers_io_util);
 
+	mutex_exit(&zpd->zpers_zfs_lock);
+
 	return (0);
 }
 
